ring/batch.go

```diff
@@ -50,32 +50,56 @@ func isHTTPStatus4xx(err error) bool {
 	return false
 }
 
-// DoBatch is a special case of DoBatchWithClientError where errors
-// containing HTTP status code 4xx are treated as client errors.
+// DoBatch is a deprecated version of DoBatchWithOptions where grpc errors containing status codes 4xx are treated as client errors.
+// Deprecated. Use DoBatchWithOptions instead.
 func DoBatch(ctx context.Context, op Operation, r ReadRing, keys []uint32, callback func(InstanceDesc, []int) error, cleanup func()) error {
-	return DoBatchWithClientError(ctx, op, r, keys, callback, cleanup, isHTTPStatus4xx)
+	return DoBatchWithOptions(ctx, op, r, keys, callback, DoBatchOptions{
+		Cleanup:       cleanup,
+		IsClientError: isHTTPStatus4xx,
+	})
 }
 
-// DoBatchWithClientError request against a set of keys in the ring,
-// handling replication and failures. For example if we want to write
-// N items where they may all hit different instances, and we want them
-// all replicated R ways with quorum writes, we track the relationship
-// between batch RPCs and the items within them.
-//
-// callback() is passed the instance to target, and the indexes of the keys
-// to send to that instance.
-//
-// cleanup() is always called, either on an error before starting the batches
-// or after they all finish.
+// DoBatchOptions defines options for the DoBatchWithOptions call.
+// Zero value options are valid, as well as individual zero valued fields.
+type DoBatchOptions struct {
+	// Cleanup is always called, either on an error before starting the batches or after they are all finished.
+	// If nil, a noop will be called.
+	Cleanup func()
+
+	// IsClientError classifies errors returned by `callback()` into client or server errors.
+	// See `batchTracker.record()` function for details about how errors are combined into final error returned by DoBatchWithClientError.
+	// If nil, a default implementation is used that classifies grpc errors containing status codes 4xx as client errors.
+	IsClientError func(error) bool
+
+	// Go will be used to spawn the callback goroutines, and can be used to use a worker pool like concurrency.ReusableGoroutinesPool.
+	Go func(func())
+}
+
+func (o *DoBatchOptions) setDefaults() {
```

Comment:
Could we rename this method in something like `setDefaultsIfNeeded()`.
IMO seeing `o.setDefaults()` could be interpreted as "we get `DoBatchOptions` o, and we set default values on all its fields`. WDYT?

Reply:
I can rename it but given it's an unexported method with a single usage that fits on the same screen I don't think that any naming can lead to confusion.

Reply:
WDYT about `replaceZeroValuesWithDefaults()` then?

Reply:
> WDYT about replaceZeroValuesWithDefaults() then?

That looks good.

Reply:
Renamed.